bundles/org.openhab.core.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttBrokerConnection.java

```diff
@@ -847,38 +839,44 @@ protected boolean finalizeStopAfterDisconnect(boolean v) {
     }
 
     /**
-     * Publish a message to the broker with the given QoS and retained flag.
+     * Publish a message to the broker.
      *
      * @param topic The topic
      * @param payload The message payload
-     * @param qos The quality of service for this message
-     * @param retain Set to true to retain the message on the broker
      * @param listener A listener to be notified of success or failure of the delivery.
      */
-    public void publish(String topic, byte[] payload, int qos, boolean retain, MqttActionCallback listener) {
-        MqttAsyncClient localClient = client;
-        if (localClient == null) {
-            listener.onFailure(topic, new MqttException(0));
-            return;
-        }
-        try {
-            IMqttDeliveryToken deliveryToken = localClient.publish(topic, payload, qos, retain, listener,
-                    actionCallback);
-            logger.debug("Publishing message {} to topic '{}'", deliveryToken.getMessageId(), topic);
-        } catch (org.eclipse.paho.client.mqttv3.MqttException e) {
-            listener.onFailure(topic, new MqttException(e));
-        }
+    public void publish(String topic, byte[] payload, MqttActionCallback listener) {
+        publish(topic, payload, null, null, listener);
     }
 
     /**
-     * Publish a message to the broker.
+     * Publish a message to the broker with the given QoS and retained flag.
      *
      * @param topic The topic
      * @param payload The message payload
+     * @param qos The quality of service for this message
+     * @param retain Set to true to retain the message on the broker
      * @param listener A listener to be notified of success or failure of the delivery.
      */
-    public void publish(String topic, byte[] payload, MqttActionCallback listener) {
-        publish(topic, payload, qos, retain, listener);
+    public void publish(String topic, byte[] payload, @Nullable Integer qos, @Nullable Boolean retain, MqttActionCallback listener) {
+        final MqttAsyncClientWrapper client = this.client;
+        if (client == null) {
+            listener.onFailure(topic, new MqttException(new Throwable()));
+            return;
+        }
+
+        // use defaults when not provided by caller
+        qos = qos == null ? this.qos : qos;
+        retain = retain == null ? this.retain : retain;
```

Comment:
You are using deprecated class fields in this non-deprecated method.

Reply:
No way to remin backward compatible otherwise. When we remove the retain-setter/getter, we can default that to false (what is the default now, if not explicitely set otherwise).